### ISSUE TITLE
Different rotation conversion with OpenPCDet

### ISSUE THREAD

**Issue by XuyangBai** [open-mmlab/mmdetection3d#443]
Hi, it seems you are using a different conversion of rotation around z axis?

https://github.com/open-mmlab/mmdetection3d/blob/4002f06cc0e42e79528a04992fa35847b971077c/mmdet3d/core/bbox/structures/lidar_box3d.py#L130-L134

while in PCDet
```
    cosa = torch.cos(angle)
    sina = torch.sin(angle)
    zeros = angle.new_zeros(points.shape[0])
    ones = angle.new_ones(points.shape[0])
    rot_matrix = torch.stack((
        cosa,  sina, zeros,
        -sina, cosa, zeros,
        zeros, zeros, ones
    ), dim=1).view(-1, 3, 3).float()
    points_rot = torch.matmul(points[:, :, 0:3], rot_matrix)
```

So PCDet is counterclockwise and mmdet3d is clockwise? Am i correct? Any reason why you use clockwise angle?

**Comment by Wuziyi616** [comment]
Yes, you are right. Currently we perform clockwise rotation. This is mainly a legacy issue and we have discussed this in previous issues (see #410 #315). A refactor of the coordinate systems to resolve this issue is ongoing. However, due to the large amount of code related to this issue, we are not sure when it will be done. Sorry for the inconvenience.

**Comment by XuyangBai** [comment]
Hi @Wuziyi616 Still thanks for this great codebase. Does that mean in mmdet3d, we have to use `-yaw` as the parameter to rotate the boxes back to align with the axis since the yaw angle is counterclockwise?

Besides, are there any things I need to be careful when using this coordinate system on my own data? Did you convert the conversion of all the datasets (KITTI, waymo, nuscenes) into yours (x-front, y-left, z-up, and yaw angle 0 is x-positive axis, counterclockwise)?

**Comment by THU17cyz** [comment]
> Hi @Wuziyi616 Still thanks for this great codebase. Does that mean in mmdet3d, we have to use `-yaw` as the parameter to rotate the boxes back to align with the axis since the yaw angle is counterclockwise?

To your first question, yes, as long as you use the same rotation functions as you did when you rotate the boxes in data preparation.

To your second one, unfortunately our coordinate system currently has some problems, and the data of different datasets do not follow the same coordinate system. There are actually three coordinate systems, depth, lidar, and camera, see [here](https://github.com/open-mmlab/mmdetection3d/tree/master/mmdet3d/core/bbox/structures). I would recommend you to use the lidar coord system, which actually follows the convention [here](https://github.com/traveller59/second.pytorch#concepts).

**Comment by THU17cyz** [comment]
A refactor is going on to fix the legacy issues and add more documents to explain the coordinate systems.

**Comment by XuyangBai** [comment]
thanks a lot, looking forward to the refactor

**Comment by XuyangBai** [comment]
Sorry for reopening this issue. Just a double check: is the yaw angle of Lidar coordinate system clockwise or counterclockwise? 

https://github.com/open-mmlab/mmdetection3d/blob/237cebffd80a25b46d0ff76eff9925d800c997d0/mmdet3d/core/bbox/structures/lidar_box3d.py#L13-L32

In here, it seems the yaw angle is counterclockwise. But for the following function you rotate clockwise to get the corners? 
https://github.com/open-mmlab/mmdetection3d/blob/237cebffd80a25b46d0ff76eff9925d800c997d0/mmdet3d/core/bbox/structures/lidar_box3d.py#L81-L83

**Comment by THU17cyz** [comment]
Hi @XuyangBai ,

We actually follow the coordinate system [here](https://github.com/traveller59/second.pytorch#concepts), which means the docstring is wrong and the yaw angle is clockwise (viewed from above).

We are very sorry about the confusion caused by this and therefore a refactor is ongoing.

**Comment by THU17cyz** [comment]
Just created a pull request to modify the docstrings of the three coordinate systems to avoid confusion.

**Comment by XuyangBai** [comment]
Thanks, it is clear now.

**Comment by Petros626** [comment]
@Wuziyi616 @yezhen17 @XuyangBai  did the codebase is now updated to counter-clockwise it`s described here,
https://mmdetection3d.readthedocs.io/en/v1.0.0rc1/tutorials/coord_sys_tutorial.html

**Comment by Petros626** [comment]
@yezhen17, @XuyangBai, @Wuziyi616 ?
